### ISSUE TITLE
`getdeviceinfolist` and `ftopen` fail on ARM linux 32 bit

### ISSUE THREAD

**Issue by ReubenHill** [Gowerlabs/LibFTD2XX.jl#24]
The `getdeviceinfolist` and `ftopen` functions do not work as expected on 32 bit ARM.

Running the example code on **64 bit windows** with a USB-RS485 Cable attached:

```Julia
julia> using LibFTD2XX

julia> devs = createdeviceinfolist()
0x00000001

julia> list, elnum = getdeviceinfolist(devs)
(LibFTD2XX.FT_DEVICE_LIST_INFO_NODE[FT_DEVICE_LIST_INFO_NODE(0x00000000, 0x00000005, 0x04036001, 0x00000000, (70, 84, 48, 52, 50, 80, 68, 49, 0, 112, 103, 127, -8, 127, 0, 0), (85, 83, 66, 45, 82, 83, 52, 56, 53, 32, 67, 97, 98, 108, 101, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, -76, -106, 108, 54, -59, 121, 0, 0, -128, 106, -73, 4, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, -106, 86, -95, 126, -8, 127, 0, 0), Ptr{Nothing} @0x0000000000000000)], 0x00000001)

julia> description = String(list[1].description)
"USB-RS485 Cable"

julia> ftopen(0)
FT_HANDLE(Ptr{Nothing} @0x0000000000da7950)

```

on **ARM linux 32 bit** (rasberry pi 3)

```Julia
julia> using LibFTD2XX

julia> devs = createdeviceinfolist()
0x00000001

julia> list, elnum = getdeviceinfolist(devs)
(LibFTD2XX.FT_DEVICE_LIST_INFO_NODE[FT_DEVICE_LIST_INFO_NODE(0x00000001, 0x00000003, 0x00000000, 0x00000000, (0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0), (0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0), Ptr{Nothing} @0x00000000)], 0x00000001)

julia> description = String(list[1].description)
""

julia> ftopen(0)
ERROR: FT_DEVICE_NOT_FOUND::FT_STATUS_ENUM = 2
Stacktrace:
 [1] ftopen(::Int32) at home/reuben/.julia/packages/LibFTD2XX/src/LibFTD2XX.jl:127
 [2] top-level scope at none:0
```

It is not clear whether this is a platform of wordsize issue. Could this be something to do with the use of, e.g. `Int` anywhere in the wrapper?

EDIT: Updated stacktrace for linux

**Comment by ReubenHill** [comment]
It's worth noting that the documentation I used for reference when writing the wrappers used windows wordsize conventions throughout (e.g. `DWORD`). Perhaps there is some undocumented difference or difference I haven't found the documentation for when calling, e.g. `FT_GetDeviceInfoList` on 32 bit linux?

**Comment by ReubenHill** [comment]
Also @samuelpowell maybe check #22 is definitely installing the correct libraries?

**Comment by samuelpowell** [comment]
You say this error is on 32 bit ARM (Linux)?

```Julia
julia> ftopen(0)
ERROR: FT_DEVICE_NOT_FOUND::FT_STATUS_ENUM = 2
Stacktrace:
 [1] ftopen(::Int32) at C:\Users\Reuben\.julia\v0.6\LibFTD2XX\src\LibFTD2XX.jl:127
 [2] top-level scope at none:0`
```

Looks a lot like Windows!?

**Comment by ReubenHill** [comment]
Ha! I had to take the windows stack trace and change the values (easy copy paste access). I was pretty careful, especially with the output word length but clearly I forgot to do that for the `ftopen` error! I'll check again today and check I didn't make other mistakes in the copy paste. The problem is real though.

**Comment by samuelpowell** [comment]
You say this error is on 32 bit ARM (Linux)?

```Julia
julia> ftopen(0)
ERROR: FT_DEVICE_NOT_FOUND::FT_STATUS_ENUM = 2
Stacktrace:
 [1] ftopen(::Int32) at C:\Users\Reuben\.julia\v0.6\LibFTD2XX\src\LibFTD2XX.jl:127
 [2] top-level scope at none:0
```

**Comment by samuelpowell** [comment]
The programmers manual says “This function should only be called after calling FT_CreateDeviceInfoList. If the devices connected to the system change, the device info list will not be updated until FT_CreateDeviceInfoList is called again.“ but I can’t see that we are doing this.

**Comment by samuelpowell** [comment]
The programmers manual also says that the arguments to this function should be an `LPDWORD` whereas you use `DWORD`? Is this consistent throughout? Further, I can see the odd ccall where you use, e.g. `Int` rather than `Cint`

**Comment by samuelpowell** [comment]
`dmesg | grep FTDI`?

**Comment by samuelpowell** [comment]
Assuming the above reveals the devices to be connected I believe we have to unload the VCP drivers before using D2XX on Linux:

```
sudo rmmod ftdi_sio
sudo rmmod usbserial
```

https://www.ftdichip.com/Support/Documents/AppNotes/AN_220_FTDI_Drivers_Installation_Guide_for_Linux.pdf

**Comment by ReubenHill** [comment]
> You say this error is on 32 bit ARM (Linux)?
> 
> ```julia
> julia> ftopen(0)
> ERROR: FT_DEVICE_NOT_FOUND::FT_STATUS_ENUM = 2
> Stacktrace:
>  [1] ftopen(::Int32) at C:\Users\Reuben\.julia\v0.6\LibFTD2XX\src\LibFTD2XX.jl:127
>  [2] top-level scope at none:0`
> ```
> 
> Looks a lot like Windows!?

Edited original comment to fix.

**Comment by ReubenHill** [comment]
> Assuming the above reveals the devices to be connected I believe we have to unload the VCP drivers before using D2XX on Linux:
> 
> ```
> sudo rmmod ftdi_sio
> sudo rmmod usbserial
> ```
> 
> https://www.ftdichip.com/Support/Documents/AppNotes/AN_220_FTDI_Drivers_Installation_Guide_for_Linux.pdf

Unloading VCP seems to make no difference:
```
reuben@raspberrypi:~ $ dmesg | grep FTDI
[    2.860243] usb 1-1.1.3: Manufacturer: FTDI
[    4.677762] usbserial: USB Serial support registered for FTDI USB Serial Device
[    4.678462] ftdi_sio 1-1.1.3:1.0: FTDI USB Serial Device converter detected
[    4.690053] usb 1-1.1.3: FTDI USB Serial Device converter now attached to ttyUSB0
[   73.289229] usbserial: USB Serial deregistering driver FTDI USB Serial Device
[   73.289473] ftdi_sio ttyUSB0: FTDI USB Serial Device converter now disconnected from ttyUSB0
[  324.259005] usb 1-1.1.3: Manufacturer: FTDI
[  324.310176] usbserial: USB Serial support registered for FTDI USB Serial Device
[  324.311089] ftdi_sio 1-1.1.3:1.0: FTDI USB Serial Device converter detected
[  324.312490] usb 1-1.1.3: FTDI USB Serial Device converter now attached to ttyUSB0
[  380.496408] usbserial: USB Serial deregistering driver FTDI USB Serial Device
[  380.496817] ftdi_sio ttyUSB0: FTDI USB Serial Device converter now disconnected from ttyUSB0
[  595.109000] usb 1-1.1.3: Manufacturer: FTDI
[  595.160841] usbserial: USB Serial support registered for FTDI USB Serial Device
[  595.161950] ftdi_sio 1-1.1.3:1.0: FTDI USB Serial Device converter detected
[  595.167904] usb 1-1.1.3: FTDI USB Serial Device converter now attached to ttyUSB0
[  610.106700] usbserial: USB Serial deregistering driver FTDI USB Serial Device
[  610.107128] ftdi_sio ttyUSB0: FTDI USB Serial Device converter now disconnected from ttyUSB0
[  758.180210] usb 1-1.1.3: Manufacturer: FTDI
[  758.208964] usbserial: USB Serial support registered for FTDI USB Serial Device
[  758.209336] ftdi_sio 1-1.1.3:1.0: FTDI USB Serial Device converter detected
[  758.210246] usb 1-1.1.3: FTDI USB Serial Device converter now attached to ttyUSB0
reuben@raspberrypi:~ $ sudo rmmod ftdi_sio
[sudo] password for reuben: 
reuben@raspberrypi:~ $ sudo rmmod usbserial
reuben@raspberrypi:~ $ dmesg | grep FTDI
[    2.860243] usb 1-1.1.3: Manufacturer: FTDI
[    4.677762] usbserial: USB Serial support registered for FTDI USB Serial Device
[    4.678462] ftdi_sio 1-1.1.3:1.0: FTDI USB Serial Device converter detected
[    4.690053] usb 1-1.1.3: FTDI USB Serial Device converter now attached to ttyUSB0
[   73.289229] usbserial: USB Serial deregistering driver FTDI USB Serial Device
[   73.289473] ftdi_sio ttyUSB0: FTDI USB Serial Device converter now disconnected from ttyUSB0
[  324.259005] usb 1-1.1.3: Manufacturer: FTDI
[  324.310176] usbserial: USB Serial support registered for FTDI USB Serial Device
[  324.311089] ftdi_sio 1-1.1.3:1.0: FTDI USB Serial Device converter detected
[  324.312490] usb 1-1.1.3: FTDI USB Serial Device converter now attached to ttyUSB0
[  380.496408] usbserial: USB Serial deregistering driver FTDI USB Serial Device
[  380.496817] ftdi_sio ttyUSB0: FTDI USB Serial Device converter now disconnected from ttyUSB0
[  595.109000] usb 1-1.1.3: Manufacturer: FTDI
[  595.160841] usbserial: USB Serial support registered for FTDI USB Serial Device
[  595.161950] ftdi_sio 1-1.1.3:1.0: FTDI USB Serial Device converter detected
[  595.167904] usb 1-1.1.3: FTDI USB Serial Device converter now attached to ttyUSB0
[  610.106700] usbserial: USB Serial deregistering driver FTDI USB Serial Device
[  610.107128] ftdi_sio ttyUSB0: FTDI USB Serial Device converter now disconnected from ttyUSB0
[  758.180210] usb 1-1.1.3: Manufacturer: FTDI
[  758.208964] usbserial: USB Serial support registered for FTDI USB Serial Device
[  758.209336] ftdi_sio 1-1.1.3:1.0: FTDI USB Serial Device converter detected
[  758.210246] usb 1-1.1.3: FTDI USB Serial Device converter now attached to ttyUSB0
[  828.397480] usbserial: USB Serial deregistering driver FTDI USB Serial Device
[  828.397728] ftdi_sio ttyUSB0: FTDI USB Serial Device converter now disconnected from ttyUSB0
reuben@raspberrypi:~ $ julia

               _
   _       _ _(_)_     |  Documentation: https://docs.julialang.org
  (_)     | (_) (_)    |
   _ _   _| |_  __ _   |  Type "?" for help, "]?" for Pkg help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 1.0.0 (2018-08-08)
 _/ |\__'_|_|_|\__'_|  |  Official https://julialang.org/ release
|__/                   |
```
```Julia
julia> using LibFTD2XX

julia> devs = createdeviceinfolist()
0x00000001

julia> list, elnum = getdeviceinfolist(devs)
(LibFTD2XX.FT_DEVICE_LIST_INFO_NODE[FT_DEVICE_LIST_INFO_NODE(0x00000001, 0x00000003, 0x00000000, 0x00000000, (0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0), (0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0), Ptr{Nothing} @0x00000000)], 0x00000001)

julia> description = String(list[1].description)
""

julia> ftopen(0)
ERROR: FT_DEVICE_NOT_FOUND::FT_STATUS_ENUM = 2
Stacktrace:
 [1] ftopen(::Int32) at /home/reuben/.julia/packages/LibFTD2XX/pglcd/src/LibFTD2XX.jl:127
 [2] top-level scope at none:0

```

**Comment by ReubenHill** [comment]
To narrow down the search: `ftopen(0)` should work immediately when any device is attached - nothing has to be called first.

**Comment by ReubenHill** [comment]
I note that `ftopen(0)` should work immediately when any device is attached - nothing has to be called first. It looks like the driver just isn't operating as expected.

Perhaps more than just 
```
sudo rmmod ftdi_sio
sudo rmmod usbserial
```

has to be done before using it on linux?

**Comment by samuelpowell** [comment]
There's nothing wrong with the code, it's just a problem with permissions (in addition to removing the other kernel modules). Fixed on test system, and instructions for fix open in #25 .
